Implement periodic cpu usage tracking in resource groups

phd3 opened the pull request:
This PR contains the following changes at a high level:

* Periodically update and cache cpu usage for resource groups.
* Add memory and cpu usage update at query completion.
* Update eligibility through a post-order traversal in `updateEligibilityAll` after resource usage update to avoid calling `updateEligibility` recursively in the `internalRefreshStats`.
* Rename `processQueuedQueries` to reflect what the method is doing
* move memory limit check in `canRunMore` for better readability.

Resolves #1059 . @dain @wagnermarkd

raghavsethi commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -166,7 +162,7 @@ public ResourceGroupInfo getFullInfo()
                     hardConcurrencyLimit,
                     maxQueuedQueries,
-                    DataSize.succinctBytes(cachedMemoryUsageBytes),
+                    DataSize.succinctBytes(cachedResourceUsage.getMemoryUsageBytes()),
```
Now that you're unifying this, I would probably add cpu usage here for consistency.

raghavsethi commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -669,24 +668,34 @@ private void startInBackground(ManagedQueryExecution query)
+
+            if (runningQueries.keySet().contains(query)) {
+                ResourceUsage cachedUsage = runningQueries.get(query);
+                // cpu usage is tracked in a cumulative sense over time, memory usage relates to a particular time instant.
```
Nit: To be consistent, capitalize CPU.

Perhaps something like this is a little clearer:
```
// CPU is measured cumulatively (i.e. total query CPU used until this moment)
// Memory is measured instantaneously (how much memory the query is using at this moment)

raghavsethi commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -669,24 +668,34 @@ private void startInBackground(ManagedQueryExecution query)
-            }
-            if (runningQueries.contains(query)) {
+
+            if (runningQueries.keySet().contains(query)) {
```
Why are we no longer checking for USER_ERROR/success? Should we allow system failures to impact user-available CPU?

phd3 replied:
In the current implementation, the user gets punished for memory usage even if the query fails due to a system error.  

Also, now since cpu usage is updated on the fly, user might already start getting punished before the query fails. 

That is why I felt that we should not discriminate between success/failure here. What do you think?

raghavsethi commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -697,36 +706,67 @@ private void queryFinished(ManagedQueryExecution query)
+     * Traverse all nodes in the subtree rooted at this node and update their eligibility. The traversal happens in
+     * a post-order fashion, i.e. eligibility of all the children is updated before the node itself.
+     */
+    protected void updateEligibilityAll()
```
This is a very confusing method name. Perhaps we could rename both this and the existing `updateEligibility` to something that reflects which direction the recursion is going (upward/downward) or (parent/children).

phd3 replied:
Updating method names to `updateEligibilityOnPathToRoot` and `updateEligibilityInSubtree`.

raghavsethi commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -858,7 +893,10 @@ private boolean canRunMore()
+            long cpuUsageMillis = cachedResourceUsage.getCpuUsageMillis();
+            long memoryUsageBytes = cachedResourceUsage.getMemoryUsageBytes();
+
+            if ((cpuUsageMillis >= hardCpuLimitMillis) || (memoryUsageBytes > softMemoryLimitBytes)) {
```
This gets me thinking about the soft vs hard concept again. I believe the way we're using CPU is here is also like a 'soft' limit, since at the hard limit we will kill the query.

phd3 replied:
A running query is killed for its cpu usage _only_ when query's cpu crosses `query.max-cpu-time`. I don't think we ever kill a running query based on the resource group usage right now. Current behavior is that overshooting `hardCpuLimitMillis` prevents all incoming queries in that group from running. Overshooting `softCpuLimitMillis` practically lowers the concurrency limit. Do you think we should change this behavior? If we do, I think that such a change will be independent to the work in this PR. LMK if I missed anything.

raghavsethi replied:
Ah no, you're right. I misremembered/misunderstood.

raghavsethi commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -921,9 +958,11 @@ public RootInternalResourceGroup(String name, BiConsumer<InternalResourceGroup,
+        public synchronized void updateGroupsAndProcessQueuedQueries()
         {
-            internalRefreshStats();
+            updateResourceUsageAndGetDelta();
```
Maybe I don't understand this, but if updateResourceUsageAndGetDelta is a recursive function and internalRefreshStats is not, then I believe this will add greatly to how CPU-intensive this call is.

phd3 replied:
`internalRefreshStats` and `updateResourceUsageAndGetDelta` are both recursive. One difference is in how they update the usage. The usage delta in `updateResourceUsageAndGetDelta` is returned from the stack through a child call.  `internalRefreshStats` stores the updated usage on the heap, the parent picks it up and does the delta calculation itself.

raghavsethi commented on presto-main/src/test/java/io/prestosql/execution/MockResourceUpdatingQueryExecution.java:
```diff
@@ -0,0 +1,139 @@
+import static io.prestosql.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class MockResourceUpdatingQueryExecution
```
Perhaps MockQueryExecutionWithResources?

phd3 replied:
Trying to see if we can merge the two mock executions as suggested by Dain.

raghavsethi commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/BenchmarkResourceGroup.java:
```diff
@@ -51,7 +51,7 @@
     public Object benchmark(BenchmarkData data)
     {
-        data.getRoot().processQueuedQueries();
+        data.getRoot().updateGroupsAndProcessQueuedQueries();
```
What are the benchmark results for large trees?

dain commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java:
```diff
@@ -321,14 +323,341 @@ public void testHardCpuLimit()
+
+        // Set the same values in all the groups for some configurations
+        List<InternalResourceGroup> groups = ImmutableList.of(root, rootChild1, rootChild2, rootChild1Child1, rootChild1Child2);
+        groups.forEach(group -> {
```
you can use `Stream.of(root, rootChild1, rootChild2, rootChild1Child1, rootChild1Child2).forEach(groups -> {`

dain commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java:
```diff
@@ -321,14 +323,341 @@ public void testHardCpuLimit()
+        q3.consumeCpuTime(new Duration(2, SECONDS));
+
+        // This invocation will update the cached usage for the nodes
+        // root: 9s, rootChild1: 7s, rootChild2:2s, rootChild1Child1: 2s, rootChild1Child2: 5s
```
Being able to validate the resource usage snapshot directly here would make this test a lot more readable.

phd3 commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java:
```diff
@@ -321,14 +323,341 @@ public void testHardCpuLimit()
+        // is useful for testing cpu update during quota generation, without the invocation of updateGroupsAndProcessQueuedQueries.
+        MockResourceUpdatingQueryExecution q3 = new MockResourceUpdatingQueryExecution();
+        child.run(q3);
+        assertEquals(q3.getState(), RUNNING);
```
Yes, I added this test to document this behavior. `refreshAndStartQueries` method is run periodically by the `InternalResourceGroupManager`. It contains invocations of `generateCpuQuota` and `updateResourceUsageAndGetDelta` methods. The case mentioned in the test here only arises if a query comes in between invocations of these two methods. This is possible since  `refreshAndStartQueries` is not synchronized.

raghavsethi replied:
It's definitely not great behavior and could lead to a lot of user confusion, did your change make this worse? If not, then you can address it in a follow up PR.

phd3 commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java:
```diff
@@ -321,14 +323,341 @@ public void testHardCpuLimit()
+        child.run(q1);
+        assertEquals(q1.getState(), RUNNING);
+
+        // The cached cpu usage will be set to 2s (> 1.5s limit) when updateGroupsAndProcessQueuedQueries is invoked
```
I agree that it is a better way to assert the status of resource group usage. But for this direct verification, we'd need to put a lock on root for creating a snapshot for `cachedResourceUsage`, since this object is guarded by it. I'm not sure if it's a good idea to use the root lock just for this test, since someone else may start using it in main code. What do you think?

phd3 commented:
@dain @raghavsethi I've updated the PR addressing your comments.

Numbers for TestResourceGroups benchmark:

Benchmark | (children) | (queries) | Time taken (us/op) (Before) | Time taken (us/op) (With `updateEligibilityInSubTree`method) | Time taken (us/op)(After removing `updateEligibilityInSubTree` method as in `c16e711 `)
-- | -- | -- | -- | -- | --
BenchmarkResourceGroup.benchmark | 1000 | 100 | 1.364 | 216.476 | 2.499
BenchmarkResourceGroup.benchmark | 1000 | 1000 | 13.428 | 236.021 | 33.543
BenchmarkResourceGroup.benchmark | 1000 | 10000 | 495.993 | 1148.303 | 1211.227
BenchmarkResourceGroup.benchmark | 10000 | 100 | 1.094 | 2353.871 | 2.405
BenchmarkResourceGroup.benchmark | 10000 | 1000 | 12.18 | 2489.451 | 26.708
BenchmarkResourceGroup.benchmark | 10000 | 10000 | 445.518 | 4522.486 | 717.43
BenchmarkResourceGroup.benchmark | 100000 | 100 | 1.217 | 33370.087 | 2.151
BenchmarkResourceGroup.benchmark | 100000 | 1000 | 14.851 | 31872.003 | 26.752
BenchmarkResourceGroup.benchmark | 100000 | 10000 | 455.631 | 32303.722 | 534.325

After seeing that the benchmark takes a lot longer to execute with `updateEligibilityInSubTree` method, I replaced its usage with `updateEligibilityOnPathToRoot` calls in `internalCpuQuotaGeneration` and `updateResourceUsageAndGetDelta`.  This change (a bit more than) doubles the time taken for the update (from before this PR), which seems reasonable to me given that we're updating two values per node now.

Removing `updateEligibilityInSubTree` requires us to invoke `updateEligibilityOnPathToRoot` method inside and `internalGenerateCpuQuota`. This works well for shallow trees, even with a high fanout. But I've a concern that it might not work well for deep trees. Do you think we should optimize this further?

raghavsethi commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -646,87 +646,116 @@ private void updateEligibility()
+            if (lastUsage != null) {
+                // CPU is measured cumulatively (i.e. total CPU used until this moment by the query). Memory is measured
+                // instantaneously (how much memory the query is using at this moment). At query completion, memory usage
+                // drops to zero. CPU usage is extracted from query execution.
```
Remove this part of the comment: `CPU usage is extracted from query execution`, it seems redundant.

raghavsethi commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -646,87 +646,116 @@ private void updateEligibility()
             }
-            updateEligibility();
+
+            // Query was already cleaned up
```
This code would be easier to understand if you did an early exit instead.

phd3 replied:
hmm, I did this refactoring to minimize the number of lookups in runningQueries and queuedQueries  ( https://github.com/prestosql/presto/pull/1128#discussion_r307529092 ). But if this is difficult to read, I can revert it.

phd3 commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -327,7 +324,7 @@ public void setSoftMemoryLimit(DataSize limit)
             this.softMemoryLimitBytes = limit.toBytes();
             if (canRunMore() != oldCanRun) {
-                updateEligibility();
+                updateEligibilityOnPathToRoot();
```
@raghavsethi I completely removed the downward method (updateEligibilityInSubtree) after looking at the benchmark results. But I still kept the name `updateEligibilityOnPathToRoot` for the `updateEligibility` method,(without any changes to its implementation) since the new name is more descriptive. So this is just a name change overall.

`InternalResourceGroup#run` method checks the limits by invoking `canRunMore` on all the nodes from leaf to root. So I think we're okay here just updating eligibility up the path to root, and not worry about the nodes below it. Are you wondering if the current implementation should be changed?

phd3 commented:
@raghavsethi I think `updateEligibilityInSubtree` method created some confusion. I've removed it [(here](https://github.com/prestosql/presto/pull/1128/commits/c16e711f5f76fc1e81f562804c6705579658468e) looking at the benchmark result. That might simplify things, and work well for shallow trees at least.  Happy to schedule a call if required for discussing this further.

raghavsethi commented:
If that's the case, it seems like we one have one `updateEligibility` method, can we rename it back?

phd3 commented:
@raghavsethi Addressed your comments.

raghavsethi commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java:
```diff
@@ -313,22 +321,390 @@ public void testHardCpuLimit()
+        // Query completion updates the cached usage to 2s. This is important because q1 will be removed
+        // from runningQueries at this point. Therefore, updateGroupsAndProcessQueuedQueries invocation will not be
+        // able to update CPU usage for it later.
+        Stream.of(root, child).forEach(group -> assertExceedsCpuLimit(group, 4));
```
Nit: We usually use `ImmutableList.of` rather than `Stream.of`

phd3 replied:
I used this based on Dain's suggestion here: https://github.com/prestosql/presto/pull/1128#discussion_r307977551

raghavsethi commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java:
```diff
@@ -313,22 +321,390 @@ public void testHardCpuLimit()
+        Stream.of(root, child).forEach(group -> assertWithinCpuLimit(group, 2));
+        assertEquals(q2.getState(), QUEUED);
+
+        // q2 starts running after groups are updated. The assertion is to make sure that CPU usage is not double counted.
```
Nit: These comments are helpful but could be less verbose. For example: `q2 should run after groups are updated. CPU usage should not be double counted`

raghavsethi commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java:
```diff
@@ -313,22 +321,390 @@ public void testHardCpuLimit()
+        RootInternalResourceGroup root = new RootInternalResourceGroup("root", (group, export) -> {}, directExecutor());
+        InternalResourceGroup child = root.getOrCreateSubGroup("child");
+
+        // Set the common configurations for the two groups
```
Redundant comment.

raghavsethi commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java:
```diff
@@ -313,22 +321,390 @@ public void testHardCpuLimit()
+
+        root.setHardCpuLimit(new Duration(16, MILLISECONDS));
+        rootChild1.setHardCpuLimit(new Duration(6, MILLISECONDS));
+        // Setting a higher limit for leaf nodes to make sure they are always in the limit
```
Minor nit: I would reorder these so that the leafs are the last lines in this block. This is because I think of full-line comments as usually applying to a block of code.

phd3 replied:
the 3 nodes below the comment are leaves indeed. I added an extra space to separate them from the intermediate ones. I'm not sure if you meant something else.

raghavsethi commented on presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java:
```diff
@@ -313,22 +321,390 @@ public void testHardCpuLimit()
+        // q6 starts running, because usage in rootChild2 and root are within their limits.
+        assertEquals(q6.getState(), RUNNING);
+
+        // q5 starts running after rootChild1's usage comes within the limit
```
Excellent testing!

raghavsethi commented:
Minor changes requested, looks generally good.

phd3 commented on presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java:
```diff
@@ -669,64 +672,93 @@ private void startInBackground(ManagedQueryExecution query)
+
+            ResourceUsage lastUsage = runningQueries.get(query);
+            // Query is running
+            if (lastUsage != null) {
```
changed the comment

phd3 commented:
@raghavsethi Thanks for the review, I've updated the PR addressing your latest comments.

> It's definitely not great behavior and could lead to a lot of user confusion, did your change make this worse? If not, then you can address it in a follow up PR.

This has been the behavior before the change as well. But it starves a queued query only if new queries submitted between the invocation of cpu quota generation and updateGroupsAndProcessQueuedQueries exceed limits instantaneously.

raghavsethi commented:
Can you go ahead and squash your commits into a single commit?

raghavsethi commented:
@phd3 could you provide a draft of the release notes for this change?

raghavsethi commented:
Merged, thanks for your contribution!

ImTangYun commented:
when we upgrade to version 318 or later we found that there is a very big bug

raghavsethi commented:
Could you provide your resource groups config file?
